File: src/main/recursion/Subsets.java
import java.util.ArrayList;
import java.util.List;

public class Subsets {

    public static void main(String[] args) {
        findSubsets(3, new ArrayList<>());
    }

    private static void findSubsets(int n, ArrayList<Integer> subset) {

        if (n == 0) {
            printSubset(subset);
            return;
        }

        // add
        subset.add(n);
        findSubsets(n - 1, subset);

        // not add
        subset.remove(subset.size() - 1);
        findSubsets(n - 1, subset);
    }

    private static void printSubset(List<Integer> subset) {
        for (Integer integer : subset) {
            System.out.print(integer + " ");
        }
        System.out.println();
    }

}

File: src/main/recursion/Subsets.kt
// Print all the subsets of a set of first n natural numbers

fun main() {
    findSubsets(3, ArrayList())
}

private fun findSubsets(n: Int, subset: MutableList<Int>) {

    if (n == 0) {
        printSubset(subset)
        return
    }

    // add
    subset.add(n)
    findSubsets(n - 1, subset)

    // not add
    subset.remove((subset.size - 1))
    findSubsets(n - 1, subset)
}

private fun printSubset(subset: List<Int>) {
    subset.forEach {
        print("$it ")
    }
    println()
}